src/cmd/flux-cancel.py

```diff
@@ -0,0 +1,187 @@
+##############################################################
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+
+import argparse
+import logging
+import os
+import pwd
+import sys
+from operator import itemgetter
+
+import flux
+from flux.job import cancel_async
+
+LOGGER = logging.getLogger("flux-cancel")
+
+
+def parse_args():
+    description = """
+    Cancel pending or running jobs with an optional exception note.
+    """
+    parser = argparse.ArgumentParser(
+        prog="flux-cancel",
+        usage="flux cancel [OPTIONS] [JOBID...]",
+        description=description,
+        formatter_class=flux.util.help_formatter(),
+    )
+    parser.add_argument("--all", action="store_true", help="Target all jobs")
+    parser.add_argument(
+        "-n", "--dry-run", action="store_true", help="Do not cancel any jobs"
+    )
+    parser.add_argument(
+        "-q",
+        "--quiet",
+        action="store_true",
+        help="Suppress output when no jobs match",
+    )
+    parser.add_argument(
+        "-u",
+        "--user",
+        type=str,
+        metavar="USER",
+        help="Set target user or 'all' (instance owner only)",
+    )
+    parser.add_argument(
+        "-S",
+        "--states",
+        action="append",
+        help="List of job states to target (default=active)",
+    )
+    parser.add_argument(
+        "-m",
+        "--message",
+        type=str,
+        metavar="NOTE",
+        help="Set optional cancel exception note",
+    )
+    parser.add_argument(
+        "jobids",
+        metavar="JOBID...",
+        type=flux.job.JobID,
+        nargs="*",
+        help="jobids to cancel",
+    )
+    return parser.parse_args()
+
+
+def log(msg):
+    """Log to stderr without logging INFO prefix"""
+    print(f"flux-cancel: {msg}", file=sys.stderr)
+
+
+def cancel(args):
+    if not args.jobids:
+        raise ValueError("No target jobs were provided")
+    if args.dry_run:
+        count = len(args.jobids)
+        log(f"Would cancel {count} job{'s'[:count^1]}")
+        sys.exit(0)
+    h = flux.Flux()
+    futures = {x: cancel_async(h, x, args.message) for x in args.jobids}
+    errors = 0
+    for jobid, future in futures.items():
+        try:
+            future.get()
+        except OSError as exc:
+            errors += 1
+            LOGGER.error(f"{jobid}: {exc.strerror}")
+    if errors > 0:
+        sys.exit(1)
+
+
+def cancelall(args):
+
+    STATES = {
```

Comment:
This seems like logic that might be needed elsewhere - any reason to one-off define it here, or any reason to not define it in flux.constants? If it's just used here this should be OK - but if we find we are reproducing this elsewhere (which could lead to bugs) lets consider a single place to import from.